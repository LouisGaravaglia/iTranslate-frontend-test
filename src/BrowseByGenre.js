import React,  {useState, useRef, useEffect, useCallback} from 'react';
import {useSpring, animated} from 'react-spring';
import './Sass/App.scss';
//COMPONENT IMPORTS
import LyricsTranslation from "./LyricsTranslation";
import Tracks from "./Tracks";
import LanguageSelect from "./LanguageSelect";
import Artists from "./Artists";
import Albums from "./Albums";
import Genres from "./Genres";
import Categories from "./BrowseCategories";
import FlashMessage from "./FlashMessage";
import ToTopArrow from "./ToTopArrow";
//REDUX IMPORTS
import {useDispatch, useSelector} from "react-redux";
import {getGenres} from "./actionCreators/BrowseRoute/Genre/getGenresCreator";
import {resetLanguageError, resetTranslationError, resetLyricsError, resetGeneralError} from "./actionCreators/handleErrorsCreator";
//CUSTOM HOOK IMPORTS
import useOnScreen from "./hooks/useOnScreen";
//IONICONS IMPORTS
import IosMusicalNotes from 'react-ionicons/lib/IosMusicalNotes';

function BrowseByGenre() {
  //STATE FOR ANIMATIONS
  const [bgColor, setBgColor] = useState("#4e1eff");
  //REDUX STORE
  const dispatch = useDispatch();
  const genres = useSelector(store => store.genres);
  const lyrics = useSelector(store => store.lyrics);
  const artists = useSelector(store => store.artists);
  const albums = useSelector(store => store.albums);
  const tracks = useSelector(store => store.tracks);
  const selectedTrackId = useSelector(store => store.selectedTrack.trackId);
  const translation = useSelector(store => store.translation);
  const translationError = useSelector(store => store.errors.translationError);
  const languageError = useSelector(store => store.errors.languageError);
  const lyricsError = useSelector(store => store.errors.lyricsError);
  const generalError = useSelector(store => store.errors.generalError);
  //REFS FOR PAGE TRAVERSAL
  const categoryRef = useRef();
  const selectGenresRef = useRef();
  const artistsResultsRef = useRef();
  const albumResultsRef = useRef();
  const selectLanguageRef = useRef();
  const trackResultsRef = useRef();
  const showLyricsTranslationRef = useRef();
  //STATE FOR FLASH MESSAGES
  const [noLyricsFlashMessage, setNoLyricsFlashMessage] = useState(false);
  const [languageNotFoundFlashMessage, setLanguageNotFoundFlashMessage] = useState(false);
  const [translationErrorFlashMessage, setTranslationErrorFlashMessage] = useState(false);
  const [generalErrorFlashMessage, setGeneralErrorFlashMessage] = useState(false);

////////////////////////////////////////////////////  USE EFFECTS  ////////////////////////////////////////////////////

  //LISTENS FOR ANY CHANGES IN ERRORS IN STATE AND WILL TRIGGER FLASH MESSAGES ACCORDINGLY
  useEffect(() => {

    const displayFlashMessage = () => {

        if (lyricsError) {
          setNoLyricsFlashMessage(true);
          dispatch(resetLyricsError());
        };
        if (languageError) {
          setLanguageNotFoundFlashMessage(true);
          dispatch(resetLanguageError());
        };
        if (translationError) {
          setTranslationErrorFlashMessage(true);
          dispatch(resetTranslationError());
        };
        if (generalError) {
          setGeneralErrorFlashMessage(true);
          dispatch(resetGeneralError());
        };
    };
    displayFlashMessage();
  }, [languageError, translationError, lyricsError, generalError, dispatch]);

  //GET ALL GENRES IN DB AND STORE THEM FOR THE BROWSE BY GENRE COMPONENT
  useEffect(() => {

    async function getSeedData() {
      dispatch(getGenres());
    };
    getSeedData();
  }, [dispatch]);


  //SKIP OVER THE CATEGORIES SINCE THE USER NEEDED TO SEE THAT IN ORDER TO GET TO THIS COMPONENT
  useEffect(() => {

    const scrollPastCategories = () => {
      selectGenresRef.current.scrollIntoView({behavior: "smooth"});
    };
    scrollPastCategories();
  }, []);

  //FUNCTION TO BE CALLED IN BELOW USE-EFFECTS TO SCROLL TO NEXT DIV AFTER CLICK
  const scrollToNextDiv = useCallback(async (state, ref) => {

    if (state && state !== "Could not read language value") {
      ref.current.scrollIntoView({behavior: "smooth"});
    };
  }, []);

  //SCROLL DOWN TO SHOW ARTISTS WHEN GENRE HAS BEEN SELECTED AND CORRESPONDING ARTISTS SET IN STATE
  useEffect(() => {scrollToNextDiv(artists, artistsResultsRef);}, [artists, artistsResultsRef, scrollToNextDiv]);

  //SCROLL DOWN TO ALBUMS RESULTS DIV WHEN ALBUMS ARE SET IN STATE
  useEffect(() => {scrollToNextDiv(albums, albumResultsRef);}, [albums, albumResultsRef, scrollToNextDiv]);

  //SCROLL DOWN TO TRACKS DIV WHEN TRACKS HAS BEEN SELECTED AND SET IN STATE
  useEffect(() => {scrollToNextDiv(tracks, trackResultsRef);}, [tracks, trackResultsRef, scrollToNextDiv]);

  //SCROLL DOWN TO LANGUAGE SEARCH BAR WHEN SELECTED TRACK HAS BE SET IN STATE
  useEffect(() => {scrollToNextDiv(lyrics, selectLanguageRef);}, [lyrics, selectLanguageRef, scrollToNextDiv]);

  //SCROLL DOWN TO SHOW LYRICS AND TRANSLATION WHEN TRANSLATION HAS BE SET IN STATE
  useEffect(() => {scrollToNextDiv(translation, showLyricsTranslationRef);}, [translation, showLyricsTranslationRef, scrollToNextDiv]);

////////////////////////////////////////////////////  ANIMATION FOR BACKGROUND COLOR  ////////////////////////////////////////////////////

  const categoriesInView = useOnScreen(categoryRef, {threshold: 0.7});
  const genresInView = useOnScreen(selectGenresRef, {threshold: 0.7});
  const artistsInView = useOnScreen(artistsResultsRef, {threshold: 0.7});
  const albumsInView = useOnScreen(albumResultsRef, {threshold: 0.7});
  const selectLanguageInView = useOnScreen(selectLanguageRef, {threshold: 0.7});
  const trackResultsInView = useOnScreen(trackResultsRef, {threshold: 0.7});
  const lyricsTranslationInvView = useOnScreen(showLyricsTranslationRef, {threshold: 0.7});

  useEffect(() => {
    const changeInView = (selectLanguageInView, albumsInView, trackResultsInView, lyricsTranslationInvView, artistsInView, categoriesInView, genresInView) => {

      if (categoriesInView) {
        setBgColor("#4e1eff");
      } else if (genresInView) {
        setBgColor("#6e3eee");
      } else if (artistsInView) {
        setBgColor("#9161db");
      } else if (albumsInView) {
        setBgColor("#aa79ce");
      } else if (trackResultsInView) {
        setBgColor("#c594bf");
      } else if (selectLanguageInView) {
        setBgColor("#e3b2af");
      } else if (lyricsTranslationInvView) {
        setBgColor("#fecda1");
      };
    };
  changeInView(selectLanguageInView, albumsInView, trackResultsInView, lyricsTranslationInvView, artistsInView, categoriesInView, genresInView);
  }, [selectLanguageInView, albumsInView, trackResultsInView, lyricsTranslationInvView, artistsInView, categoriesInView, genresInView]);

  const springProps = useSpring({
    backgroundColor: bgColor,
    config: {duration: 300}
  });

////////////////////////////////////////////////////  CLICK EVENTS  ////////////////////////////////////////////////////

  const scrollToGenres = () => {
    selectGenresRef.current.scrollIntoView({behavior: "smooth"});
  };

////////////////////////////////////////////////////  JSX VARIABLES  ////////////////////////////////////////////////////

  //DISPLAY THE THREE CATEGORIES
  const ChooseCategoryDiv = (
    <animated.div onClick={scrollToGenres} style={springProps} ref={categoryRef}>
      <Categories needAnimation={false}/>
    </animated.div>
  );

  //DISPLAY GENRES
  let selectGenresDiv;
  
  if (genres) selectGenresDiv = (
     <animated.div style={springProps} ref={selectGenresRef}>
       <Genres />
     </animated.div>
  );

  //DISPLAY MUSIC ICON IF THERE IS AN ERROR LOADING GENRES
  if (!genres) selectGenresDiv = (
    <animated.div style={springProps} ref={selectGenresRef}>
      <div className="Main-Container">
        <div className="Loading-Box">
          <IosMusicalNotes fontSize="300px" color="orange"/>
        </div>
      </div>
    </animated.div>
  );

  //DISPLAY ARTISTS FROM SELECTED GENRE
  let ArtistsResultsDiv;
  
  if (artists) ArtistsResultsDiv = (
    <animated.div style={springProps} ref={artistsResultsRef}>
      <Artists typeOfArtists="genre"/>
    </animated.div>
  );

  //DISPLAY ALBUMS FROM SELECTED ARTIST
  let AlbumResultsDiv;
  
  if (albums) AlbumResultsDiv = (
    <animated.div style={springProps} ref={albumResultsRef}>
      <Albums typeOfAlbums="genre"/>
    </animated.div>
  );

  //DISPLAY TRACKS FROM SELECTED ALBUM
  let TrackResultsDiv;



    if (tracks) TrackResultsDiv = (
        <animated.div style={springProps} ref={trackResultsRef}>
          <Tracks results={tracks} typeOfResults={"tracks"} itemsPerPage={1} animateIn={true} typeOfTracks="genre"/>
        </animated.div>
      );

      if (!tracks) TrackResultsDiv = (
        <animated.div style={springProps} ref={trackResultsRef}>
        </animated.div>
      );

 
  //DISPLAY LANGUAGE SELECTION SEARCH BAR
  let LanguageSelectDiv;

  if (lyrics) LanguageSelectDiv = (
    <animated.div style={springProps} ref={selectLanguageRef}>
      <LanguageSelect selectedTrackId={selectedTrackId} typeOfSearch="genre-language"/>
    </animated.div>
  );

  //DISPLAY LYRICS AND TRANSLATION
  let LyricsTranslationDiv;
  
  if (translation && translation !== "Could not read language value") LyricsTranslationDiv = (
    <animated.div style={springProps}>
      <div className="inViewPlaceholder" ref={showLyricsTranslationRef}></div>
      <ToTopArrow topRef={categoryRef} topInView={categoriesInView}/>
      <LyricsTranslation  typeOfLyricsTranslation="genre"/>
    </animated.div>
  );

////////////////////////////////////////////////////  RETURN  ////////////////////////////////////////////////////

  return (
    <>
      <div className="Flash-Messages-Container">
        {noLyricsFlashMessage && (<FlashMessage setState={setNoLyricsFlashMessage} message="Unfortunately there are no Lyrics for that song yet."/> )}
        {languageNotFoundFlashMessage && (<FlashMessage setState={setLanguageNotFoundFlashMessage} message="That Language was not found, please try again."/> )}
        {translationErrorFlashMessage && (<FlashMessage setState={setTranslationErrorFlashMessage} message="Sorry, we couldn't get a translation at this moment."/> )}
        {generalErrorFlashMessage && (<FlashMessage setState={setGeneralErrorFlashMessage} message="Uh oh, something went wrong. Please try again."/> )}
      </div>
      {ChooseCategoryDiv}
      {selectGenresDiv}
      {ArtistsResultsDiv}
      {AlbumResultsDiv}
      {TrackResultsDiv}
      {LanguageSelectDiv}
      {LyricsTranslationDiv}
    </>
  );
};

export default BrowseByGenre;