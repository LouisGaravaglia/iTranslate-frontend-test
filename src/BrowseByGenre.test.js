import React from 'react'
import {render, cleanup} from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import {Provider} from 'react-redux'
import {createStore} from "redux";
import BrowseByGenre from "./BrowseByGenre";

//CREATE A MOCK INTERSECTION OBSERVER CLASS FOR TESTING AND APPEND IT TO THE WINDOW OBJECT
const mockIntersectionObserver = class {
  constructor() {}
  observe() {}
  unobserve() {}
  disconnect() {}
};

beforeEach(() => {
  window.IntersectionObserver = mockIntersectionObserver;
});

afterEach(cleanup);

const startingState = {
  selectedTrack: {trackName: "chanel", trackId: 42}, 
  errors: {translationError: false, languageError: false, lyricsError: false, searchError: false, generalError: false}
};

/**
 * reducer function to mimic the reducer used in Redux.
 * @param {object} state - holds an object of data used within the components
 * @param {object} action - object where the values are strings used to fire certain actions
 */
function reducer(state = startingState, action) {
  switch (action.type) {
    case "ADD_TRACKS":
      return {...state, tracks: action.tracks};
    default:
      return state;
  };
};

/**
 * Function that renders the component within a redux environment since most 
 * components rely on redux for data in order to mount.
 * @param {ReactComponent} component - private spotify client id
 * @param {function} store - creates the react store using the createStore method imported from redux
 */
function renderWithRedux(
  component,
  {initialState, store = createStore(reducer, initialState)} = {}
) {
  return {
    ...render(<Provider store={store}>{component}</Provider>)
  };
};

//SMOKE TEST
describe('Smoke Test for BrowseByGenre component', () => {

  //NEED TO ESTABLISH THAT SCROLLINTOVIEW IS A FUNCTION IN ORDER FOR TESTS TO PASS
  window.HTMLElement.prototype.scrollIntoView = function() {};

  it('renders without crashing', () => {
    renderWithRedux(<BrowseByGenre />);
  });
});

//SNAPSHOT TEST
describe('Snapshot Test for BrowseByGenre component', () => {

  it('matches snapshot', () => {
    const {asFragment} = renderWithRedux(<BrowseByGenre />);
    expect(asFragment()).toMatchSnapshot();
  });
});