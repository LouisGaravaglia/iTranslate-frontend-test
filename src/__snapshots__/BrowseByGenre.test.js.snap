// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Test for BrowseByGenre component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Flash-Messages-Container"
  />
  <div
    style="background-color: rgb(78, 30, 255);"
  >
    <div
      class="Browse-Landing"
    >
      <span
        class="Hover-Box"
        style="display: inline-block; transform: scale(1);"
      >
        <p
          class="Browse-Categories"
        >
          Artists
        </p>
      </span>
      <span
        class="Hover-Box"
        style="display: inline-block; transform: scale(1);"
      >
        <p
          class="Browse-Categories"
        >
          Genre
        </p>
      </span>
      <span
        class="Hover-Box"
        style="display: inline-block; transform: scale(1);"
      >
        <p
          class="Browse-Categories"
        >
          Danceability
        </p>
      </span>
    </div>
  </div>
  <div
    style="background-color: rgb(78, 30, 255);"
  >
    <div
      class="Main-Container"
    >
      <div
        class="Loading-Box"
      >
        <svg
          class="sc-bdVaJa fUuvxv"
          fill="orange"
          height="300px"
          rotate="0"
          viewBox="0 0 1024 1024"
          width="300px"
        >
          <path
            d="M812.6 96.4c-9.4 1.8-404 78.4-412.4 80s-16.2 7.2-16.2 16c0 0 0 477 0 480.2s-0.2 14.4-4.8 23.4c-6.2 11.8-17 20.4-32.2 25.4-6.6 2.2-15.6 4.2-26.2 6.6-48.2 10.8-128.8 29.2-128.8 103.6 0 62.2 44.8 90.2 83.4 95 4.2 0.6 9 1.4 14.2 1.4 0 0 0 0 0 0 13.4 0 48-2.6 78.4-22.4 22-14.4 48.2-42.8 48.2-95.6v-438l384-78c0 0 0 273.2 0 281.4s-0.4 17.8-5 26.8c-6.2 11.8-17 20.4-32.4 25.4-6.6 2.2-15.6 4.2-26.2 6.6-48.2 10.8-128.8 29-128.8 103.4 0 67.4 52.6 91.2 83.6 94.6 2.4 0.2 5.2 0.2 8.2 0.2 0 0 0 0 0 0 20.4 0 51.4-5 77.6-20 35.8-20.6 55-53.6 55-96.4v-604.2c-0.2-8.8-7.6-17.8-19.6-15.4z"
          />
        </svg>
      </div>
    </div>
  </div>
  <div
    style="background-color: rgb(78, 30, 255);"
  />
</DocumentFragment>
`;
